src/bottom_up_iterator.jl

```diff
@@ -0,0 +1,223 @@
+"""
+	abstract type BottomUpIterator <: ProgramIterator
+
+Enumerates programs in a bottom-up fashion. This means that it starts with the smallest programs and gradually builds up to larger programs.
+The exploration of the search space is done by making use of the priority function, which associates each program with its cost.
+
+Concrete implementations of this iterator should implement the following methods:
+- `order(iter::BottomUpIterator, grammar::ContextSensitiveGrammar)::Vector{Int64}`: Returns the order in which the rules should be enumerated.
+- `pick(iter::BottomUpIterator, grammar::ContextSensitiveGrammar, state::BottomUpState, rule::Int64)::Vector{RuleNode}`: Returns the programs that can be created by applying the given rule.
```

Comment:
Instead of the return type being `Vector{X}`, we could make it an `Iterable` or `Iterator`. that would include `Vector` but also a proper iterator that could save memory